### ISSUE TITLE
Banning for loops when a TS for(...of...) loop might be better

### ISSUE THREAD

**Issue by chrismbarr** [palantir/tslint#1335]
Basically a standard `for` loop should only be allowed if the incrementer variable is used in the block of the loop for anything other than accessing the current array item.

 👍 **This is Fine** since the array index is needed for something
```ts
for (var j = 0; j <= arr.length; j++) {
    doMath(j);
}
```

👎 **This could be replaced with a `for(...of...)` loop** since the index is only used to access the array
```ts
for (var i = 0; i <= arr.length; i++) {
    doThing(arr[i]);
}
```
**Here's the preferred way to write the above loop**
```ts
for (var item of arr) {
    doThing(item);
}
```

----

Is this a feasible rule idea?  I've [already begun some work on it](https://github.com/chrismbarr/tslint/commit/78fa93f101eb622feab6e2f1dc5655140a9e0a12), but I'll work on it more later.  Right now it's only partially working since I'm not sure how to look for variable usages within a certain scope/Block/SyntaxList.  Any tips would be appreciated!

**Comment by adidahiya** [comment]
I'm a fan of this rule; it could be called `prefer-for-of-loop`.

> I'm not sure how to look for variable usages within a certain scope/Block/SyntaxList

One idea is to use the regular `languageService.getDocumentHighlights` method, iterate through the highlights and verify that the source code positions of those spans are within the Block you are searching within.

**Comment by scameron** [comment]
If you are adding elements to the end of the array being iterated inside the loop, don’t you need to use a standard for?

**Comment by beenotung** [comment]
I agree for-of is more readable and easier to maintain, however, it may hurt the performance.

For example, in Firefox: for loop with index has 219,705 ops/sec, while for-of has only 35,827 ops/sec. Which is 84% slower

demo:
https://jsperf.com/foreach/19

**Comment by wortwart** [comment]
This rule should not be default because in the DOM there are array-like objects which you can't iterate with for..of or .forEach() - at least not without extra hassle. I've been struggling repeatedly with this rule in node collections like you get with document.querySelectorAll(). 
Classic for loops are highly performant, simple, and readable and should IMHO be considered good practice.
